src/server/bg_services/log_replication_scanner.js

```diff
@@ -122,28 +122,36 @@ class LogReplicationScanner {
     async process_candidates(src_bucket, dst_bucket, candidates) {
         const src_dst_objects_list = await this.head_objects(src_bucket.name, dst_bucket.name, candidates);
 
+        const copy_keys = [];
+        const delete_keys = [];
+
         for (const candidate of src_dst_objects_list) {
-            await this.process_candidate(src_bucket, dst_bucket, candidate);
+            const action = await this.process_candidate(src_bucket, dst_bucket, candidate);
+            if (action === 'copy') {
+                copy_keys.push(candidate.key);
+            } else if (action === 'delete') {
+                delete_keys.push(candidate.key);
+            }
         }
+
+        this.copy_objects(src_bucket, dst_bucket, copy_keys);
+        this.delete_objects(dst_bucket, delete_keys);
     }
 
     async process_candidate(src_bucket, dst_bucket, candidate) {
-        // each candidate represents several action over the same key
-        // that are sorted by time.
-        const items = candidate.data || [];
+        // each candidate represents latest action per key
+        const item = candidate.data && candidate.data[0];
```

Comment:
Not sure but here `item` can be `undefined` as well right? It might happen if we return an empty array from `create_candidates`?

Reply:
That looks right.